bindata/assets/kube-apiserver/pod.yaml

```
@@ -19,16 +19,44 @@ spec:
-        done
+
+        LOCK=/var/log/kube-apiserver/.lock
+        echo "Acquiring exclusive lock ${LOCK} ..."
+
+        # Waiting for 130s max for old kube-apiserver's watch-termination process to exit and remove the lock.
+        # Two cases:
+        # 1. if kubelet does not start the old and new in parallel (i.e. works as expected), the flock will always succeed without any time.
+        # 2. if kubelet does overlap old and new pods for up to 130s, the flock will wait and immediate return when the old finishes.
+        #
+        # NOTE: We can increase 130s for a bigger expected overlap. But a higher value means less noise about the broken kubelet behaviour, i.e. we hide a bug.
+        # NOTE: Do not tweak these timings without considering the livenessProbe initialDelaySeconds
+        exec {LOCK_FD}>${LOCK} && flock --verbose -w 130 "${LOCK_FD}" || {
+          echo "$(date -Iseconds -u) kubelet did not terminate old kube-apiserver before new one" >> /var/log/kube-apiserver/lock.log
+          echo -n ": WARNING: kubelet did not terminate old kube-apiserver before new one."
+
+          # We failed to acquire exclusive lock, which means there is old kube-apiserver running in system.
+          # We don't want to fail here and we will continue, risking audit log corruption and termination log corruption.
+          # Since we utilize SO_REUSEPORT, we need to make sure the old kube-apiserver stopped listening.
+          #
+          # NOTE: This is a fallback for broken kubelet, if you observe this please report a bug.
+          echo -n "Waiting for port 6443 to be released due to likely bug in kubelet or CRI-O "
+          while [ -n "$(ss -Htan state listening '( sport = 6443 )')" ]; do
+            echo -n "."
+            sleep 1
+            (( tries += 1 ))
+            if [[ "${tries}" -gt 60 ]]; then
+              echo "Timed out waiting for port :6443 to be released, this is likely a bug in kubelet or CRI-O"
+              break
```

so we continue and risk the audit logs?

---

yes, no change from previous version

---

i can put comment there

---

I would like to fail hard here.